fxa-oauth-server/lib/routes/key_data.js

```diff
@@ -48,40 +50,47 @@ module.exports = {
     const requestedClientId = req.payload.client_id;
 
     return P.all([
-      verify(req.payload.assertion),
+      verifyAssertion(req.payload.assertion),
       db.getClient(Buffer.from(requestedClientId, 'hex')).then((client) => {
         if (client) {
           // find all requested scopes that are allowed for this client.
           const allowedScopes = ScopeSet.fromString(client.allowedScopes || '');
           const scopeLookups = requestedScopes.filtered(allowedScopes).getScopeValues().map(s => db.getScope(s));
           return P.all(scopeLookups).then((result) => {
-            return result.filter((s) => !! (s && s.hasScopedKeys));
+            return result.filter((s) => !! (s && s.hasScopedKeys))
+              .map(s => {
+                // When we implement key rotation these values will come from the db.
+                // For now all scoped keys have the default values.
+                s.keyRotationSecret = DEFAULT_KEY_ROTATION_SECRET;
+                s.keyRotationTimestamp = DEFAULT_KEY_ROTATION_TIMESTAMP;
+                return s;
+              });
           });
         } else {
           logger.debug('keyDataRoute.clientNotFound', { id: req.payload.client_id });
           throw AppError.unknownClient(requestedClientId);
         }
       })
-    ]).then((results => {
-      logger.debug('keyDataRoute.results', JSON.stringify(results));
-      const assertionData = results[0];
-      const scopeResults = results[1];
+    ]).spread((claims, scopes) => {
+      const iat = claims.iat || claims['fxa-lastAuthAt'];
       const response = {};
-
-      if (assertionData['fxa-lastAuthAt'] < (Date.now() - AUTH_EXPIRES_AFTER_MS) / 1000) {
-        throw AppError.staleAuthAt(assertionData['fxa-lastAuthAt']);
-      }
-
-      scopeResults.forEach((keyScope) => {
+      scopes.forEach((keyScope) => {
+        const keyRotationTimestamp = Math.max(claims['fxa-generation'], keyScope.keyRotationTimestamp);
+        // If the assertion certificate was issued prior to a key-rotation event,
+        // we don't want to revel the new secrets to such stale assertions,
+        // even if they are technically still valid.
+        if (iat < (keyRotationTimestamp / 1000)) {
+          throw AppError.staleAuthAt(iat);
+        }
```

Comment:
On the other hand, this is the check that we *would* need if we actually ever tried to do a rotation of a specific scoped key.